troubleshoot-lock-conflicts.md

```diff
@@ -0,0 +1,234 @@
+---
+title: TiDB Lock Conflict Handling
+summary: How to analyze lock conflicts and solutions for that.
+category: troubleshoot
+---
+
+# TiDB Lock Conflict Handling
+ 
+TiDB can support complete distributed transactions. As of version of v3.0, TiDB provides optimistic transaction mode and pessimistic transaction mode. This document introduces how to troubleshooting and resolve when TiDB has usually lock conflict.
+
+## Optimistic Transaction Mode
+
+In the transaction of TiDB which has two-phase commit (2PC) that included the Prewrite phase and the Commit phase. The procedure is as follows:
+
+![two-phase of optimistic transaction mode](/media/troubleshooting-lock-pic-01.png)
+ 
+More than details information about Percolator and TiDB’s algorithm of the transaction, see them in [Google's Percolator](https://ai.google/research/pubs/pub36726)。
+
+### Prewrite Pharse
+ 
+As a 2PC of the Prewrite phase, TiDB will add primary lock and secondary lock to target keys. At the too much add locks request from clients, TiDB will feedback such as write conflict or the key has been locked error to the TiDB log and client. There are some details errors and treatment suggestion below a part of introduced.
+ 
+#### Read-write Conflict
+ 
+As TiDB Server receives a read data request from a client, that will get unique of a timestamp is start_ts at the physical time. When the transaction is waiting for reading data of the start_ts that is later than the key of the commit_ts version.  Because the latest transaction knows which is add a lock of the other transactions on this key,  but it doesn't know which is a phase of this transaction between the Prewrite phase and the Commit phase.The procedure is as follows:
+
+![read-write conflict](/media/troubleshooting-lock-pic-04.png)
+
+If Txn0 of a transaction has been completed for the Prewrite Phase to the target key, and then which is the status of the Commit Phase right now. At the same time, the Txn1 is reading to the same target key. Txn1 needs to read start_ts which is later than the commit_ts of the latest version of the target key. When the Txn1’s  start_ts is later than the Txn0's lock_ts, the Txn1 must be waiting for the target key 's lock has been cleanup, but it hasn’t done. As a result, The Txn1 can’t confirm the Txn0 has been committed or not. That scene is a read-write conflict between the Txn1 and the Txn0.
```

Comment:
```suggestion
Txn0 completes the Prewrite phase and enters the Commit phase. At this time, Txn1 requests to read the same target key. Txn1 needs to read the target key of the latest commit_ts that is smaller than its start_ts. Because Txn1’s start_ts is smaller than Txn0's lock_ts, Txn1 must wait for the target key's lock to be cleaned, but it hasn’t been done. As a result, Txn1 cannot confirm whether Txn0 has been committed or not. Thus, a read-write conflict between Txn1 and Txn0 happens.
```
In my understanding, Txn0 means transaction 0. So Txn0 of a transaction -> Txn0